rumqttd/src/protocol/v5/publish.rs

```diff
@@ -34,6 +35,10 @@ pub fn read(
     bytes.advance(variable_header_index);
     let topic = read_mqtt_bytes(&mut bytes)?;
 
+    if !valid_topic(from_utf8(&topic).map_err(|_| Error::TopicNotUtf8)?) {
```

Comment:
Here the validation is done (again) when obtaining `len`. This can be a problem when `len` is used during encoding since the utf8 check is not cheap at all. 

Have you thought about a `TryFrom` impl for all the packet types in here? Validate the parameters once at construction time (and maybe precalculate the length once.